src/Language/PureScript/Sugar/BindingGroups.hs

```diff
@@ -224,21 +225,36 @@ toBindingGroup moduleName (CyclicSCC ds') = do
 
 toDataBindingGroup
   :: MonadError MultipleErrors m
-  => SCC Declaration
+  => SCC (Declaration, (ProperName 'TypeName, VertexType), [(ProperName 'TypeName, VertexType)])
   -> m Declaration
-toDataBindingGroup (AcyclicSCC d) = return d
-toDataBindingGroup (CyclicSCC [d]) = case isTypeSynonym d of
+toDataBindingGroup (AcyclicSCC (d, _, _)) = return d
+toDataBindingGroup (CyclicSCC [(d, _, _)]) = case isTypeSynonym d of
   Just pn -> throwError . errorMessage' (declSourceSpan d) $ CycleInTypeSynonym (Just pn)
   _ -> return d
 toDataBindingGroup (CyclicSCC ds')
-  | all (isJust . isTypeSynonym) ds' = throwError . errorMessage' (declSourceSpan (head ds')) $ CycleInTypeSynonym Nothing
-  | kds@((ss, _):_) <- concatMap kindDecl ds' = throwError . errorMessage' ss . CycleInKindDeclaration $ fmap snd kds
-  | otherwise = return . DataBindingGroupDeclaration $ NEL.fromList ds'
+  | kds@((ss, _):_) <- concatMap (kindDecl . getDecl) ds' = throwError . errorMessage' ss . CycleInKindDeclaration $ fmap snd kds
+  | isNothing . depsTerminate [] [] $ getName <$> ds' = throwError . errorMessage' (declSourceSpan (getDecl (head ds'))) $ CycleInTypeSynonym Nothing
+  | otherwise = return . DataBindingGroupDeclaration . NEL.fromList $ getDecl <$> ds'
   where
   kindDecl (KindDeclaration sa _ pn _) = [(fst sa, Qualified Nothing pn)]
   kindDecl (ExternDataDeclaration sa pn _) = [(fst sa, Qualified Nothing pn)]
   kindDecl _ = []
 
+  getDecl (decl, _, _) = decl
+  getName (_, name, _) = name
+  lookupVert name = find ((==) name . getName) ds'
+
+  depsTerminate seen _ [] = Just seen
```

Comment:
I think this function could do with some documentation. Would this be accurate?

> This function ensures that all paths through an SCC which start and end at a given type synonym involve at least one vertex which is not a type synonym. This is necessary to prevent loops while expanding type synonyms.
> In this function:
> - `seen` is the list of names we know are safe, either because they are not type synonyms, or because we have already checked that all paths through the SCC starting and ending at that vertex involve one non-type-synonym vertex
> - `stop` is the current path of unbroken type synonym vertices being checked for cycles
>
> This function either returns Nothing in the case that a type synonym cycle was found, or Just a list of seen names in the case where no cycles were found.

Also, could we have it return a `Left stop` if we did find a cycle, and then use `stop` in the error message, in the same way that the other CycleInX error constructors take a list of names involved in the cycle?

Reply:
I've changed the implementation to use SCC to find all cycles in group, and I've changed the error to include all the names in the cycles.